sns/parity_math.c

```diff
@@ -194,254 +419,1239 @@ static void (*fidx_recover[M0_PARITY_CAL_ALGO_NR])(struct m0_parity_math *math,
+		return M0_ERR(-EDQUOT);
+
+	ir_failure_register(ir, failed_index);
+
+	return M0_RC(0);
+}
+
+M0_INTERNAL int m0_sns_ir_mat_compute(struct m0_sns_ir *ir)
+{
+	int ret = 0;
+
+	M0_PRE(ir != NULL);
+
+	ret = ir_mat_compute(ir);
+	return (ret == 0) ? M0_RC(ret) : M0_ERR(ret);
+}
+
+M0_INTERNAL void m0_sns_ir_fini(struct m0_sns_ir *ir)
+{
+	uint32_t i;
+
+	M0_ENTRY("ir=%p", ir);
+	M0_PRE(ir != NULL);
+
+	for (i = 0; i < ir_blocks_count(ir); ++i)
+		if (ir->si_blocks[i].sib_bitmap.b_words != NULL)
+			m0_bitmap_fini(&ir->si_blocks[i].sib_bitmap);
+
+	ir_rs_fini(ir);
+
+	m0_free(ir->si_blocks);
+	M0_LEAVE();
+}
+
+M0_INTERNAL int m0_sns_ir_recover(struct m0_sns_ir *ir,
+				  struct m0_bufvec *bufvec,
+				  const struct m0_bitmap *bitmap,
+				  uint32_t failed_index,
+				  enum m0_sns_ir_block_type block_type)
+{
+	size_t                  block_idx = 0;
+	size_t                  b_set_nr;
+	struct m0_sns_ir_block *blocks;
+	struct m0_sns_ir_block *alive_block;
+	int                     ret;
+
+	M0_ENTRY("ir=%p, bufvec=%p, bitmap=%p, failed_index=%u, block_type=%u",
+		 ir, bufvec, bitmap, failed_index, block_type);
+
+	M0_PRE(ir != NULL && bufvec != NULL && bitmap != NULL);
+
+	b_set_nr = m0_bitmap_set_nr(bitmap);
+	M0_PRE(b_set_nr > 0);
+	M0_PRE(ergo(b_set_nr > 1, block_type == M0_SI_BLOCK_REMOTE));
+	M0_PRE(ergo(block_type == M0_SI_BLOCK_REMOTE,
+		    is_valid_block_idx(ir, failed_index)));
+	M0_PRE(ergo(block_type == M0_SI_BLOCK_REMOTE,
+	            ir->si_blocks[failed_index].sib_status ==
+		    M0_SI_BLOCK_FAILED));
+	M0_PRE(ergo(block_type == M0_SI_BLOCK_LOCAL, ir->si_local_nr != 0));
+	if (b_set_nr == 1) {
+		for (block_idx = 0; block_idx < bitmap->b_nr; ++block_idx) {
+			if (m0_bitmap_get(bitmap, block_idx))
+				break;
+		}
+	}
+
+	VALUE_ASSERT_INFO(is_valid_block_idx(ir, block_idx), block_idx);
+
+	blocks = ir->si_blocks;
+
+	switch (block_type) {
+	/* Input block is assumed to be an untransformed block, and is used for
+	 * recovering all failed blocks */
+	case M0_SI_BLOCK_LOCAL:
+
+		M0_CNT_DEC(ir->si_local_nr);
+		alive_block = &blocks[block_idx];
+		alive_block->sib_addr = bufvec;
+
+		ret = ir_recover(ir, alive_block);
+		if (ret != 0)
+			return 	M0_ERR_INFO(ret, "Incremental recovery failed.");
+		break;
+
+	/* Input block is assumed to be a transformed block, and is
+	 * cumulatively added to a block with index failed_index. */
+	case M0_SI_BLOCK_REMOTE:
+		if (!is_usable(ir, (struct m0_bitmap*) bitmap,
+			       &blocks[failed_index]))
+			break;
+		gfaxpy(blocks[failed_index].sib_addr, bufvec,
+		       1);
+#if !ISAL_ENCODE_ENABLED
+		dependency_bitmap_update(&blocks[failed_index],
+					 bitmap);
+		if (is_data(ir, failed_index) && are_failures_mixed(ir) &&
+		    ir->si_local_nr != 0)
+			forward_rectification(ir, bufvec, failed_index);
+#endif /* !ISAL_ENCODE_ENABLED */
+		break;
+	}
+
+	return M0_RC(0);
+}
+
+/* Parity Math Helper Functions */
+
+/* m0_parity_* are to much eclectic. just more simple names. */
```

Comment:
"too much...". Please start sentences with a capital letter.

Reply:
This was present in the original code. I have updated the comment.